File: report/src/main/java/com/hartwig/actin/report/interpretation/AggregatedEvidenceFactory.java
package com.hartwig.actin.report.interpretation;

import java.util.List;

import com.google.common.collect.Lists;
import com.hartwig.actin.molecular.datamodel.MolecularRecord;
import com.hartwig.actin.molecular.datamodel.characteristics.MolecularCharacteristics;
import com.hartwig.actin.molecular.datamodel.driver.Amplification;
import com.hartwig.actin.molecular.datamodel.driver.Disruption;
import com.hartwig.actin.molecular.datamodel.driver.Fusion;
import com.hartwig.actin.molecular.datamodel.driver.HomozygousDisruption;
import com.hartwig.actin.molecular.datamodel.driver.Loss;
import com.hartwig.actin.molecular.datamodel.driver.MolecularDrivers;
import com.hartwig.actin.molecular.datamodel.driver.Variant;
import com.hartwig.actin.molecular.datamodel.driver.Virus;
import com.hartwig.actin.molecular.datamodel.evidence.ActionableEvidence;
import com.hartwig.actin.molecular.util.MolecularEventFactory;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class AggregatedEvidenceFactory {

    private static final Logger LOGGER = LogManager.getLogger(AggregatedEvidenceFactory.class);

    private AggregatedEvidenceFactory() {
    }

    @NotNull
    public static AggregatedEvidence create(@NotNull MolecularRecord molecular) {
        List<AggregatedEvidence> evidences = Lists.newArrayList();

        evidences.addAll(aggregateCharacteristicsEvidence(molecular.characteristics()));
        evidences.addAll(aggregateDriverEvidence(molecular.drivers()));

        ImmutableAggregatedEvidence.Builder aggregateBuilder = ImmutableAggregatedEvidence.builder();
        for (AggregatedEvidence evidence : evidences) {
            aggregateBuilder.putAllApprovedTreatmentsPerEvent(evidence.approvedTreatmentsPerEvent());
            aggregateBuilder.putAllExternalEligibleTrialsPerEvent(evidence.externalEligibleTrialsPerEvent());
            aggregateBuilder.putAllOnLabelExperimentalTreatmentsPerEvent(evidence.onLabelExperimentalTreatmentsPerEvent());
            aggregateBuilder.putAllOffLabelExperimentalTreatmentsPerEvent(evidence.offLabelExperimentalTreatmentsPerEvent());
            aggregateBuilder.putAllPreClinicalTreatmentsPerEvent(evidence.preClinicalTreatmentsPerEvent());
            aggregateBuilder.knownResistantTreatmentsPerEvent(evidence.knownResistantTreatmentsPerEvent());
            aggregateBuilder.suspectResistanceTreatmentsPerEvent(evidence.suspectResistanceTreatmentsPerEvent());
        }
        return aggregateBuilder.build();
    }

    @NotNull
    private static List<AggregatedEvidence> aggregateCharacteristicsEvidence(@NotNull MolecularCharacteristics characteristics) {
        List<AggregatedEvidence> evidences = Lists.newArrayList();

        if (hasCharacteristic(characteristics.isMicrosatelliteUnstable())) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.MICROSATELLITE_UNSTABLE, characteristics.microsatelliteEvidence()));
        } else if (hasEvidence(characteristics.microsatelliteEvidence())) {
            LOGGER.warn("There is evidence for microsatellite stability without presence of signature");
        }

        if (hasCharacteristic(characteristics.isHomologousRepairDeficient())) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.HOMOLOGOUS_REPAIR_DEFICIENT,
                    characteristics.homologousRepairDeficiencyEvidence()));
        } else if (hasEvidence(characteristics.homologousRepairDeficiencyEvidence())) {
            LOGGER.warn("There is evidence for homologous repair deficiency without presence of signature");
        }

        if (hasCharacteristic(characteristics.hasHighTumorMutationalBurden())) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.HIGH_TUMOR_MUTATIONAL_BURDEN,
                    characteristics.tumorMutationalBurdenEvidence()));
        } else if (hasEvidence(characteristics.tumorMutationalBurdenEvidence())) {
            LOGGER.warn("There is evidence for high tumor mutational burden without presence of signature");
        }

        if (hasCharacteristic(characteristics.hasHighTumorMutationalLoad())) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.HIGH_TUMOR_MUTATIONAL_LOAD,
                    characteristics.tumorMutationalLoadEvidence()));
        } else if (hasEvidence(characteristics.tumorMutationalLoadEvidence())) {
            LOGGER.warn("There is evidence for high tumor mutational load without presence of signature");
        }

        return evidences;
    }

    private static boolean hasCharacteristic(@Nullable Boolean characteristic) {
        return characteristic != null && characteristic;
    }

    @NotNull
    private static List<AggregatedEvidence> aggregateDriverEvidence(@NotNull MolecularDrivers drivers) {
        List<AggregatedEvidence> evidences = Lists.newArrayList();
        for (Variant variant : drivers.variants()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(variant), variant.evidence()));
        }

        for (Amplification amplification : drivers.amplifications()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(amplification), amplification.evidence()));
        }

        for (Loss loss : drivers.losses()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(loss), loss.evidence()));
        }

        for (HomozygousDisruption homozygousDisruption : drivers.homozygousDisruptions()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(homozygousDisruption), homozygousDisruption.evidence()));
        }

        for (Disruption disruption : drivers.disruptions()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(disruption), disruption.evidence()));
        }

        for (Fusion fusion : drivers.fusions()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(fusion), fusion.evidence()));
        }

        for (Virus virus : drivers.viruses()) {
            evidences.add(createAggregateEvidence(MolecularEventFactory.event(virus), virus.evidence()));
        }

        return evidences;
    }

    private static boolean hasEvidence(@Nullable ActionableEvidence evidence) {
        if (evidence == null) {
            return false;
        }

        return !evidence.approvedTreatments().isEmpty() || !evidence.externalEligibleTrials().isEmpty()
                || !evidence.onLabelExperimentalTreatments().isEmpty() || !evidence.offLabelExperimentalTreatments().isEmpty()
                || !evidence.preClinicalTreatments().isEmpty() || !evidence.knownResistantTreatments().isEmpty()
                || !evidence.suspectResistantTreatments().isEmpty();
    }

    @NotNull
    private static AggregatedEvidence createAggregateEvidence(@NotNull String event, @Nullable ActionableEvidence evidence) {
        ImmutableAggregatedEvidence.Builder builder = ImmutableAggregatedEvidence.builder();
        if (evidence == null) {
            return builder.build();
        }

        if (!evidence.approvedTreatments().isEmpty()) {
            builder.putAllApprovedTreatmentsPerEvent(event, evidence.approvedTreatments());
        }

        if (!evidence.externalEligibleTrials().isEmpty()) {
            builder.putAllExternalEligibleTrialsPerEvent(event, evidence.externalEligibleTrials());
        }

        if (!evidence.onLabelExperimentalTreatments().isEmpty()) {
            builder.putAllOnLabelExperimentalTreatmentsPerEvent(event, evidence.onLabelExperimentalTreatments());
        }

        if (!evidence.offLabelExperimentalTreatments().isEmpty()) {
            builder.putAllOffLabelExperimentalTreatmentsPerEvent(event, evidence.offLabelExperimentalTreatments());
        }

        if (!evidence.preClinicalTreatments().isEmpty()) {
            builder.putAllPreClinicalTreatmentsPerEvent(event, evidence.preClinicalTreatments());
        }

        if (!evidence.knownResistantTreatments().isEmpty()) {
            builder.putAllKnownResistantTreatmentsPerEvent(event, evidence.knownResistantTreatments());
        }

        if (!evidence.suspectResistantTreatments().isEmpty()) {
            builder.putAllSuspectResistanceTreatmentsPerEvent(event, evidence.suspectResistantTreatments());
        }

        return builder.build();
    }
}

File: common/src/main/java/com/hartwig/actin/molecular/interpretation/AggregatedEvidenceFactory.kt
package com.hartwig.actin.molecular.interpretation

import com.hartwig.actin.molecular.datamodel.MolecularRecord
import com.hartwig.actin.molecular.datamodel.characteristics.MolecularCharacteristics
import com.hartwig.actin.molecular.datamodel.driver.MolecularDrivers
import com.hartwig.actin.molecular.datamodel.evidence.ActionableEvidence
import com.hartwig.actin.molecular.util.MolecularCharacteristicEvents
import org.apache.logging.log4j.LogManager

object AggregatedEvidenceFactory {

    private val LOGGER = LogManager.getLogger(AggregatedEvidenceFactory::class.java)

    fun create(molecular: MolecularRecord): AggregatedEvidence {
        return mergeAggregatedEvidenceList(
            aggregateCharacteristicsEvidence(molecular.characteristics) + aggregateDriverEvidence(molecular.drivers)
        )
    }

    private fun aggregateCharacteristicsEvidence(characteristics: MolecularCharacteristics): List<AggregatedEvidence> {
        return listOfNotNull(
            aggregatedEvidenceForCharacteristic(
                characteristics.isMicrosatelliteUnstable,
                MolecularCharacteristicEvents.MICROSATELLITE_UNSTABLE,
                characteristics.microsatelliteEvidence,
                "microsatellite stability"
            ),
            aggregatedEvidenceForCharacteristic(
                characteristics.isHomologousRepairDeficient,
                MolecularCharacteristicEvents.HOMOLOGOUS_REPAIR_DEFICIENT,
                characteristics.homologousRepairEvidence,
                "homologous repair deficiency"
            ),
            aggregatedEvidenceForCharacteristic(
                characteristics.hasHighTumorMutationalBurden,
                MolecularCharacteristicEvents.HIGH_TUMOR_MUTATIONAL_BURDEN,
                characteristics.tumorMutationalBurdenEvidence,
                "high tumor mutational burden"
            ),
            aggregatedEvidenceForCharacteristic(
                characteristics.hasHighTumorMutationalLoad,
                MolecularCharacteristicEvents.HIGH_TUMOR_MUTATIONAL_LOAD,
                characteristics.tumorMutationalLoadEvidence,
                "high tumor mutational load"
            )
        )
    }

    private fun aggregatedEvidenceForCharacteristic(
        characteristic: Boolean?, event: String, evidence: ActionableEvidence?, characteristicName: String
    ): AggregatedEvidence? {
        if (characteristic == true) {
            return createAggregatedEvidence(event, evidence)
        } else if (hasEvidence(evidence)) {
            LOGGER.warn("There is evidence for $characteristicName without presence of signature")
        }
        return null
    }

    private fun hasEvidence(evidence: ActionableEvidence?): Boolean {
        return if (evidence == null) false else {
            listOf(
                evidence.approvedTreatments,
                evidence.externalEligibleTrials,
                evidence.onLabelExperimentalTreatments,
                evidence.offLabelExperimentalTreatments,
                evidence.preClinicalTreatments,
                evidence.knownResistantTreatments,
                evidence.suspectResistantTreatments
            ).any(Set<String>::isNotEmpty)
        }
    }

    private fun aggregateDriverEvidence(drivers: MolecularDrivers): List<AggregatedEvidence> {
        return listOf(
            drivers.variants, drivers.copyNumbers, drivers.homozygousDisruptions, drivers.disruptions, drivers.fusions, drivers.viruses
        ).flatMap { driverSet -> driverSet.map { createAggregatedEvidence(it.event, it.evidence) } }
    }

    private fun createAggregatedEvidence(event: String, evidence: ActionableEvidence?): AggregatedEvidence {
        return if (evidence == null) {
            AggregatedEvidence()
        } else {
            AggregatedEvidence(
                approvedTreatmentsPerEvent = evidenceMap(event, evidence.approvedTreatments),
                externalEligibleTrialsPerEvent = evidenceMap(event, evidence.externalEligibleTrials),
                onLabelExperimentalTreatmentsPerEvent = evidenceMap(event, evidence.onLabelExperimentalTreatments),
                offLabelExperimentalTreatmentsPerEvent = evidenceMap(event, evidence.offLabelExperimentalTreatments),
                preClinicalTreatmentsPerEvent = evidenceMap(event, evidence.preClinicalTreatments),
                knownResistantTreatmentsPerEvent = evidenceMap(event, evidence.knownResistantTreatments),
                suspectResistantTreatmentsPerEvent = evidenceMap(event, evidence.suspectResistantTreatments)
            )
        }
    }

    private fun mergeAggregatedEvidenceList(aggregatedEvidenceList: List<AggregatedEvidence>): AggregatedEvidence {
        return AggregatedEvidence(
            approvedTreatmentsPerEvent = mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::approvedTreatmentsPerEvent)),
            externalEligibleTrialsPerEvent =
            mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::externalEligibleTrialsPerEvent)),
            onLabelExperimentalTreatmentsPerEvent =
            mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::onLabelExperimentalTreatmentsPerEvent)),
            offLabelExperimentalTreatmentsPerEvent =
            mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::offLabelExperimentalTreatmentsPerEvent)),
            preClinicalTreatmentsPerEvent =
            mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::preClinicalTreatmentsPerEvent)),
            knownResistantTreatmentsPerEvent =
            mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::knownResistantTreatmentsPerEvent)),
            suspectResistantTreatmentsPerEvent =
            mergeMapsOfLists(aggregatedEvidenceList.map(AggregatedEvidence::suspectResistantTreatmentsPerEvent)),
        )
    }

    private fun mergeMapsOfLists(mapsOfSets: List<Map<String, List<String>>>): Map<String, List<String>> {
        return mapsOfSets
            .flatMap { it.entries }
            .groupBy({ it.key }, { it.value })
            .mapValues { it.value.flatten() }
    }

    private fun evidenceMap(event: String, evidenceSubSet: Set<String>): Map<String, List<String>> {
        return if (evidenceSubSet.isEmpty()) emptyMap() else mapOf(event to evidenceSubSet.toList())
    }
}
